cirq-core/cirq/ops/projector.py

```diff
@@ -0,0 +1,161 @@
+import itertools
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Union,
+)
+
+import numpy as np
+from scipy.sparse import coo_matrix
+
+from cirq import value
+from cirq.ops import raw_types
+
+
+def check_qids_dimension(qids):
+    """A utility to check that we only have Qubits."""
+    for qid in qids:
+        if qid.dimension != 2:
+            raise ValueError(f"Only qubits are supported, but {qid} has dimension {qid.dimension}")
+
+
+@value.value_equality
+class ProjectorString:
+    def __init__(
+        self,
+        projector_dict: Dict[raw_types.Qid, int],
+        coefficient: Union[int, float, complex] = 1,
+    ):
+        """Contructor for ProjectorString
+
+        Args:
+            projector_dict: a dictionary of Qbit to an integer specifying which vector to project
+                onto.
+            coefficient: Initial scalar coefficient. Defaults to 1.
+        """
+        check_qids_dimension(projector_dict.keys())
+        self._projector_dict = projector_dict
+        self._coefficient = complex(coefficient)
+
+    @property
+    def projector_dict(self) -> Dict[raw_types.Qid, int]:
+        return self._projector_dict
+
+    @property
+    def coefficient(self) -> complex:
+        return self._coefficient
+
+    def matrix(self, projector_qids: Optional[Iterable[raw_types.Qid]] = None) -> coo_matrix:
+        """Returns the matrix of self in computational basis of qubits.
+
+        Args:
+            projector_qids: Ordered collection of qubits that determine the subspace
+                in which the matrix representation of the ProjectorString is to
+                be computed. Qbits absent from self.qubits are acted on by
+                the identity. Defaults to the qubits of the projector_dict.
+
+        Returns:
+            A sparse matrix that is the projection in the specified basis.
+        """
+        projector_qids = self._projector_dict.keys() if projector_qids is None else projector_qids
+        check_qids_dimension(projector_qids)
+        idx_to_keep = []
+        for qid in projector_qids:
+            if qid in self._projector_dict:
+                idx_to_keep.append([self._projector_dict[qid]])
+            else:
+                idx_to_keep.append(list(range(qid.dimension)))
+
+        total_d = np.prod([qid.dimension for qid in projector_qids])
+
+        ones_idx = []
+        for idx in itertools.product(*idx_to_keep):
+            assert len(idx) == len(list(projector_qids))
+            d = total_d
+            kron_idx = 0
+            for i, qid in zip(idx, projector_qids):
+                d //= qid.dimension
+                kron_idx += i * d
+            ones_idx.append(kron_idx)
+
+        return coo_matrix(([1.0] * len(ones_idx), (ones_idx, ones_idx)), shape=(total_d, total_d))
```

Comment:
Missing multiplication by `self._coefficient.`

Reply:
Fixed and added a unit test. Thank you.